tensorboard/webapp/plugins/npmi/views/annotations_list/annotation/annotation_container.ts

```diff
@@ -51,6 +53,7 @@ import {RunColorScale} from '../../../../../types/ui';
       [showCounts]="showCounts$ | async"
       [sidebarWidth]="sidebarWidth$ | async"
       [colorScale]="runColorScale$ | async"
+      [runMaps]="runMaps$ | async"
```

Comment:
iiuc, the container has @Input bindings for `data` (an array of items, each of which has a runId), and `runMaps` (an array of `Run`s, which we use to convert runIds into run names).  In this PR, we jump through some hoops to get the actual run names:
- first getting current experiment ids
- then getting runs based on that
- then doing a linear lookup in the Runs array for a matching run (and `drawRunHintTexts` is called on every redraw call?)

I'd like to propose something which is hopefully more convenient:
- Introduce a new `getRunMap()` selector in `webapp/runs/store/runs_selectors.ts`, whose only purpose is to convert `state.runMetadata` into an ES6 Map, and return it
- `annotation_container.ts` uses the selector, and passes the `runMap` Map directly into `annotation_component`
- Instead of a linear `this.runMaps.find`, the component does `this.runMap.get(run)?.name || ''` to get the run name from the runId

With this, we skip the experimentId/route business, and have an actual Map to use, rather than an array.

wdyt?

Reply:
That was something I thought about as well, but I assume there was a reason why this was not an option. Otherwise the whole run fetching would have been much easier. Can you comment on this @stephanwlee?

Reply:
> With this, we skip the experimentId/route business, and have an actual Map to use, rather than an array.

You still need to supply it with the `experimentId` information (unless you are suggesting making `getRunsForCurrentRoute`.

Reply:
I was thinking of a `getRunMap()` that just returns all the Runs it knows about, without requiring any `experimentId`. 
 Although, I do see the reason to avoid offering the map in a selector.  Giving a map makes it easy for consumers to mistake the keys as a list of "all the runs", when they actually want only runs for the current route.

If we really don't want an experiment-agnostic `getRunMap()`, could we take another approach where
- annotation_container takes all the runIds from its `data` input prop
- calls `this.store.select(getRun, runId).pipe(take(1))` on each one, and collects the results
- assembles the runIdToName map itself, and passes that to annotation_component

?

The annotation_container already has runIds which are specific enough, so it has no need to be aware of experimentIds.

Reply:
Sorry, I still don't understand. The runs_reducer contain many data including `eid -> RunId[]` and `RunId -> RunMetadata`. Are you saying we should just expose the `RunId -> RunMetadata`?  I have no big issue with that except I don't think that change is necessary to unblock this PR.

Reply:
> Are you saying we should just expose the RunId -> RunMetadata?

Yup, that's my suggestion.  My intent is not to block the PR on that change, but rather to call out that `annotation_container` should not have to fetch experimentIds and that passing a "runIdToRun" Map seems more appropriate to pass into the component instead of an Array.

Reply:
> ...to pass into the component 

Do you mean big C Component? (i.e., Container vs. Component)? Ideally, you should pass neither array nor map but instead a UI data structure that you form in the Container into the Component with only minimal information that you require for rendering.

Reply:
In https://github.com/tensorflow/tensorboard/pull/4609, I'm proposing a new selector `getRunMap` returning a `Map<string, Run>`, which should make things simpler in this PR.

Assuming folks like the idea, the change in this PR could look like:
```ts
// annotation_container.ts
<annotation-component [runIdToRuns] = "runIdToRuns$ | async">
// ...
runIdToRuns$ = this.store.select(getRunMap);
```
```ts
// annotation_component.ts
@Input runIdToRuns;
// ...
      .text((d: string) => {
        return this.runIdToRuns.get(d)?.name || '';
      });
```

> Ideally, you should pass neither array nor map but instead a UI data structure that you form in the Container into the Component with only minimal information that you require for rendering.

That's true.  Alternatively, one might want `data: ValueData[]` to be changed to `dataWithRunDisplayNames: ValueDataWithRunDisplayNames[]`, and annotations_container only has to pass to its `annotation_component` just `data`, instead of `data` and `runNames`.  I'll leave that up to the PR author, both are fine to me.